### PR TITLE
Channel frequency rules can now be set for each contact

### PR DESCRIPTION
**Pull request by annando**
We can now define per contact if we want to see more or less posts from them in the feeds or if we don't want to see their posts at all.

### CODE REVIEW

**Review by MrPetovan — COMMENTED** [review]
What’s the difference between default visibility and always?

**Comment by annando** [comment]
In the "for you" channel, we only display every post of a contact, when we interact very often with that contact or if that option is activated for the contact. This rule is meant for contacts where you don't interact very much, but you want to see all posts anyway.

**Comment by MrPetovan** [comment]
Ok. Since these choices are static and limited, please convert the select box to radio buttons like for the account types with descriptions like the one you just gave me.

**Comment by annando** [comment]
Done (I just realized that my grammar is already in the weekend, I rely on your nitpicking skills) :grin:

File: src/Module/Contact/Profile.php [hunk 1]
```diff
@@ -419,7 +438,21 @@ protected function content(array $request = []): string
...
-		]);
+			'$channel_settings_label' => $channel_settings_label,
+			'$channel_visibility'     => [
+				'channel_visibility',
+				$this->t('Visibility of this contact in appropriate channels'),
+				$channel_visibility,
+				$this->t('Depending on the type of the channel not all posts from contacts are displayed by default. They for example need to have a certain amount of comments to be displayed. On the other hand there can be contacts who flood the channel, so you might want to see only some of their posts. Or you don\'t want to see their content at all, but you don\'t want to block or hide the contact completely.'),
+				$channel_visibilities
+			],
+			'$visibility_label'       => $this->t('Visibility of this contact in appropriate channels'),
```
Comment (MrPetovan) [inline src/Module/Contact/Profile.php]:
```suggestion
			'$visibility_label'       => $this->t('Frequency of this contact in relevant channels'),
```

File: src/Module/Contact/Profile.php [hunk 2]
```diff
@@ -419,7 +438,21 @@ protected function content(array $request = []): string
...
+			'$visibility_description' => $this->t('Depending on the type of the channel not all posts from contacts are displayed by default. They for example need to have a certain amount of comments to be displayed. On the other hand there can be contacts who flood the channel, so you might want to see only some of their posts. Or you don\'t want to see their content at all, but you don\'t want to block or hide the contact completely.'),
```
Comment (MrPetovan) [inline src/Module/Contact/Profile.php]:
```suggestion
			'$visibility_description' => $this->t("Depending on the type of the channel not all posts from this contact are displayed. By default, posts need to have a minimum amount of interactions (comments, likes) to show in your channels. On the other hand there can be contacts who flood the channel, so you might want to see only some of their posts. Or you don't want to see their content at all, but you don't want to block or hide the contact completely."),
```

File: src/Module/Contact/Profile.php [hunk 3]
```diff
@@ -336,6 +340,21 @@ protected function content(array $request = []): string
...
 
+		if (in_array($contact['network'], Protocol::FEDERATED)) {
+			$channel_settings_label = $this->t('Channel Settings');
+			$channel_visibility     = Contact\User::getChannelVisibility($contact['id'], $this->session->getLocalUserId());
+			$channel_visibilities   = [
+				Contact\User::VISIBILITY_DEFAULT => $this->t('Default visibility'),
+				Contact\User::VISIBILITY_ALWAYS  => $this->t('Display all posts of this contact'),
+				Contact\User::VISIBILITY_REDUCED => $this->t('Display only few posts'),
+				Contact\User::VISIBILITY_NEVER   => $this->t('Never display posts from this contact'),
+			];
```
Comment (MrPetovan) [inline src/Module/Contact/Profile.php]: The `$channel_visibilities` variable won't be used anymore when the `$channel_visibility` template variable is removed for being unused.

File: src/Module/Contact/Profile.php [hunk 4]
```diff
@@ -336,6 +340,21 @@ protected function content(array $request = []): string
...
+		} else {
+			$channel_settings_label = null;
+			$channel_visibility     = null;
+			$channel_visibilities   = null;
```
Comment (MrPetovan) [inline src/Module/Contact/Profile.php]: The `$channel_visibilities` variable won't be used anymore when the `$channel_visibility` template variable is removed for being unused.

File: src/Module/Contact/Profile.php [hunk 5]
```diff
@@ -419,7 +438,21 @@ protected function content(array $request = []): string
...
+				$channel_visibilities
+			],
+			'$visibility_label'       => $this->t('Visibility of this contact in appropriate channels'),
+			'$visibility_description' => $this->t('Depending on the type of the channel not all posts from contacts are displayed by default. They for example need to have a certain amount of comments to be displayed. On the other hand there can be contacts who flood the channel, so you might want to see only some of their posts. Or you don\'t want to see their content at all, but you don\'t want to block or hide the contact completely.'),
+			'$visibility_default'     => ['channel_visibility', $this->t('Default visibility'), Contact\User::VISIBILITY_DEFAULT, $this->t('When activated, posts by this contact are displayed in the "for you" channel, if either you interact often with this contact or if a post reached some level of interaction.'), $channel_visibility == Contact\User::VISIBILITY_DEFAULT],
```
Comment (MrPetovan) [inline src/Module/Contact/Profile.php]:
```suggestion
			'$visibility_default'     => ['channel_visibility', $this->t('Default frequency'), Contact\User::VISIBILITY_DEFAULT, $this->t('Posts by this contact are displayed in the "for you" channel if you interact often with this contact or if a post reached some level of interaction.'), $channel_visibility == Contact\User::VISIBILITY_DEFAULT],
```

File: src/Module/Contact/Profile.php [hunk 6]
```diff
@@ -419,7 +438,21 @@ protected function content(array $request = []): string
...
+			'$visibility_always'      => ['channel_visibility', $this->t('Display all posts of this contact'), Contact\User::VISIBILITY_ALWAYS, $this->t('If you follow this contact, then every post of this contact will appear on the "for you" channel'), $channel_visibility == Contact\User::VISIBILITY_ALWAYS],
```
Comment (MrPetovan) [inline src/Module/Contact/Profile.php]:
```suggestion
			'$visibility_always'      => ['channel_visibility', $this->t('Display all posts of this contact'), Contact\User::VISIBILITY_ALWAYS, $this->t('If you follow this contact, all their posts will appear on the "for you" channel'), $channel_visibility == Contact\User::VISIBILITY_ALWAYS],
```

Comment (MrPetovan) [inline src/Module/Contact/Profile.php]:
```suggestion
			'$visibility_always'      => ['channel_visibility', $this->t('Display all posts'), Contact\User::VISIBILITY_ALWAYS, $this->t('If you follow this contact, then every post of this contact will appear on the "for you" channel'), $channel_visibility == Contact\User::VISIBILITY_ALWAYS],
```

File: src/Module/Contact/Profile.php [hunk 7]
```diff
@@ -419,7 +438,21 @@ protected function content(array $request = []): string
...
+			'$visibility_reduced'     => ['channel_visibility', $this->t('Display only few posts'), Contact\User::VISIBILITY_REDUCED, $this->t('When a contact creates a lot of posts in a short period, this setting reduces the number of displayed posts in every channel.'), $channel_visibility == Contact\User::VISIBILITY_REDUCED],
+			'$visibility_never'       => ['channel_visibility', $this->t('Never display posts from this contact'), Contact\User::VISIBILITY_NEVER, $this->t('Posts from this contact will never be displayed in any channel'), $channel_visibility == Contact\User::VISIBILITY_NEVER],
```
Comment (MrPetovan) [inline src/Module/Contact/Profile.php]:
```suggestion
			'$visibility_never'       => ['channel_visibility', $this->t('Never display posts'), Contact\User::VISIBILITY_NEVER, $this->t('Posts from this contact will never be displayed in any channel'), $channel_visibility == Contact\User::VISIBILITY_NEVER],
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]
Now that I better understand the contact setting, I believe it should be named "frequency" instead of "visibility". We already have a "visibility" concept in Friendica shared with Mastodon regarding a post scope, either "Public" or  "Private", so I'd prefer if we picked a more accurate term for this feature.

**Comment by annando** [comment]
The changes are done.

**Comment by MrPetovan** [comment]
Please rename the whole feature from “Visibility” (already used elsewhere i  the project) to “Frequency”. This includes variables, constants, class, methods and database field names.

**Comment by annando** [comment]
I couldn't easily separate other changes to the code, so this PR now additionally contains an improvement for picking the post with the most activity when `FREQUENCY_REDUCED` is selected. Also it contains an improvement to the automatic reload for new content. Also this PR reverts the changes to the database structure update that caused a database error upon updating.

File: src/Model/Contact/User.php [hunk 8]
```diff
@@ -314,6 +318,53 @@ public static function isCollapsed(int $cid, int $uid): bool
...
+	 * @throws \ImagickException
+	 */
+	public static function getChannelFrequency(int $cid, int $uid): int
+	{
+		$cdata = Contact::getPublicAndUserContactID($cid, $uid);
+		if (empty($cdata)) {
+			return false;
+		}
+
+		$visibility = self::FREQUENCY_DEFAULT;
```
Comment (MrPetovan) [inline src/Model/Contact/User.php]:
```suggestion
		$frequency = self::FREQUENCY_DEFAULT;
```

File: src/Model/Contact/User.php [hunk 9]
```diff
@@ -314,6 +318,53 @@ public static function isCollapsed(int $cid, int $uid): bool
...
+
+		if (!empty($cdata['public'])) {
+			$public_contact = DBA::selectFirst('user-contact', ['channel-frequency'], ['cid' => $cdata['public'], 'uid' => $uid]);
+			if (DBA::isResult($public_contact)) {
+				$visibility = $public_contact['channel-frequency'] ?? self::FREQUENCY_DEFAULT;
```
Comment (MrPetovan) [inline src/Model/Contact/User.php]:
```suggestion
				$frequency = $public_contact['channel-frequency'] ?? self::FREQUENCY_DEFAULT;
```

File: src/Model/Contact/User.php [hunk 10]
```diff
@@ -314,6 +318,53 @@ public static function isCollapsed(int $cid, int $uid): bool
...
+			}
+		}
+
+		return $visibility;
```
Comment (MrPetovan) [inline src/Model/Contact/User.php]:
```suggestion
		return $frequency;
```

File: src/Model/Contact/User.php [hunk 11]
```diff
@@ -314,6 +318,53 @@ public static function isCollapsed(int $cid, int $uid): bool
...
+		$cdata = Contact::getPublicAndUserContactID($cid, $uid);
+		if (empty($cdata)) {
+			return;
+		}
+
+		DBA::update('user-contact', ['channel-frequency' => $visibility], ['cid' => $cdata['public'], 'uid' => $uid], true);
+	}
+
+	/**
+	 * Returns the channel visibility state for contact id and user id
```
Comment (MrPetovan) [inline src/Model/Contact/User.php]:
```suggestion
	 * Returns the channel frequency state for contact id and user id
```

File: src/Module/Conversation/Network.php [hunk 12]
```diff
@@ -340,14 +336,20 @@ protected function parseRequest(array $request)
...
-		} elseif (in_array($this->selectedTab, [TimelineEntity::RECEIVED, TimelineEntity::STAR])) {
+		} elseif (in_array($this->selectedTab, [TimelineEntity::RECEIVED, TimelineEntity::STAR]) || $this->timeline->isCommunity($this->selectedTab)) {
 			$this->order = 'received';
 		} elseif (($this->selectedTab == TimelineEntity::CREATED) || $this->timeline->isChannel($this->selectedTab)) {
 			$this->order = 'created';
 		} else {
 			$this->order = 'commented';
 		}
 
+		// Upon updates in the background and order by last comment we order by receive date,
```
Comment (MrPetovan) [inline src/Module/Conversation/Network.php]:
```suggestion
		// Upon updates in the background and order by last comment we order by received date,
```

File: src/Model/Contact/User.php
Comment (MrPetovan) [inline src/Model/Contact/User.php:1]: I realized that this class should actually be moved to `src/Model/User/Contact.php` to match both the database table name and the intent in English, but it doesn't need to be done in this PR.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Model/Contact/User.php (hunk 8 above)
Comment (annando) [reply src/Model/Contact/User.php]: Done

File: src/Model/Contact/User.php (hunk 9 above)
Comment (annando) [reply src/Model/Contact/User.php]: Done

File: src/Model/Contact/User.php (hunk 10 above)
Comment (annando) [reply src/Model/Contact/User.php]: Done

File: src/Model/Contact/User.php (hunk 11 above)
Comment (annando) [reply src/Model/Contact/User.php]: Done

File: src/Module/Conversation/Network.php (hunk 12 above)
Comment (annando) [reply src/Module/Conversation/Network.php]: Done

File: src/Model/Contact/User.php
Comment (annando) [reply src/Model/Contact/User.php:1]: I must admit, I want to continue with the channels.

**Review by MrPetovan — APPROVED** [review]

**Pull request merged**